woocommerce/woocommerce
# [COT] Effectively synchronize orders from the posts table to the custom orders table

### All Submissions:






### Changes proposed in this Pull Request:

#### TLDR:

This pull request combines the the settings UI/scheduled actions for orders sync work that was implemented in #32062 (and adjusted in #32706) with the posts/meta tables to custom tables migration classes that were implemented in #32034 and #32640. Additionally, creating and updating orders will now cause a synchronization process to be started, provided that synchronization is enabled. It also fixes a bug related to the automatic authoritative table switch and does some significant refactor on the migration classes.

#### More precisely:

1. The scheduled action handler implemented in the `DataSynchronizer` class will now effectively synchronize orders from the posts table to the custom orders table (**but not the other way around!** That part will be implemented separately).
2. The order creation and update process will be captured by the `DataSynchronizer` class (via `woocommerce_after_order_object_save` hook) and will start a data synchronization process (only if data synchronization is enabled in settings).
    * Note/remember that another way to start a data synchronization process is to go to _Settings - Advanced - Custom data stores_ and hitting "Save", even without actually modifying any settings.
    * Also, remember that for now, actual sync will only happen when the posts table is the authoritative data source for orders.
3. Fixes the handling of the "Switch to using the other table when sync finishes" option handling, which was reversed (also, option cleanup was happening _before_ that option was actually read which made it useless).
4. Classes in the `Database\Migrations\CustomOrderTable` namespace have been refactored:
    * Some have been renamed.
    * More descriptive doc comments have been added.
    * (Now) redundant methods used during the development of the classes have been removed.
    * Type hints for parameters and return values have been added to all the methods.
5. Changes have been implemented to ease testing:
    * The code that throws an exception if a switch of the authoritative table is attempted while there are out of sync orders has been temporarily disabled. Thus it's now possible to run `wp option update woocommerce_custom_orders_table_enabled yes|no`.
    * Saving the settings will always restart the sync process, even if it's in progress already (this is currently the only way to recover from an error thrown inside the scheduled action that performs the sync).
    * Added a new filter, `woocommerce_orders_cot_and_posts_sync_step_size`, to customize the count of orders that will be synchronized in each scheduled action run.

**NOTE:** Given the scope of this pull request it's recommended to review the individual commits separately.
   
### How to test the changes in this Pull Request:

_Prerrequisite:_ having some orders in the posts table.

1. Enable the feature by adding the following code snippet somewhere, for example at the end of `woocommerce.php`:

```php
wc_get_container()
   ->get('Automattic\WooCommerce\Internal\DataStores\Orders\CustomOrdersTableController')
   ->show_feature();
```

2. If you hadn't created the orders table previously, go to _Status - Tools_ and run the  "Create and  fill custom orders table" tool. If you had, make sure that the table is empty (`truncate table wp_wc_orders` is fine).

3. Go to _WooCommerce - Settings - Advanced - Custom data stores_, you should see a message saying that there are out of sync orders (all your orders, actually) but no mention to sync being in progress.

4. Make sure that the "Use the WordPress table" radio is selected and that "Keep the posts table and the orders table synchronized" is checked, and save (save even if you didn't need to change anything).

5. Once the page reloads, you'll see that the message has changed to indicate that the sync process is in progress. Reload again repeatedly, until the message disappears.

7. Verify that the `wp_wc_orders` table now contains proper entries for all the existing orders.

8. Repeat the process (manually truncate the table, save settings with "Keep the posts table and the orders table synchronized" checked) but this time, once the first reload of the settings page check "Switch to using the orders table as the authoritative data store when sync finishes" and save. Now, after the sync finishes you should see that the "Use the WooCommerce orders table" is selected instead.

9. Make the posts table authoritative again.

10. Modify any of the existing orders, for example change its status.

11. Immediately after the previous step reload the settings page. You should see that the message appears again saying that there's one order pending sync, _**and**_ that the sync is in progress.

12. Once the sync has happened verify that the corresponding order (same value for the `id` field) has been updated accordingly in `wp_wc_orders`.

### Other information:




### FOR PR REVIEWER ONLY:
